pkg/ebpf/bpf_error.go

```diff
@@ -0,0 +1,207 @@
+	padding uint32
+	ret     int64 // return value
+	id      BPFErrorType
+	cpu     uint32                 // cpu number
+	line    uint32                 // line number
+	file    [BPFMaxErrFileLen]byte // filename
+}
+
+func (b BPFError) Error() string {
+	if bpfErr, found := errorMap[BPFErrorType(b.id)]; found {
+		return bpfErr
+	}
+
+	return errorMap[BPFErrUnspec]
+}
+
+func (b BPFError) Count() uint32 {
+	return b.count
+}
+
+func (b BPFError) Return() int64 {
+	return int64(b.ret)
+}
+
+func (b BPFError) ID() uint32 {
+	return uint32(b.id)
+}
+
+func (b BPFError) Type() BPFErrorType {
+	return BPFErrorType(b.id)
+}
+
+func (b BPFError) CPU() uint32 {
+	return b.cpu
+}
+
+func (b BPFError) Line() uint32 {
+	return b.line
+}
+
+func (b BPFError) File() []byte {
+	return b.file[:]
+}
+
+func (b BPFError) FileAsString() string {
+	errFile := string(b.file[:])
+	errFileEnd := strings.Index(errFile, "\x00")
+
+	return errFile[:errFileEnd]
+}
+
+func (b BPFError) Size() int {
+	return int(unsafe.Sizeof(b.count)) +
+		int(unsafe.Sizeof(b.padding)) +
+		int(unsafe.Sizeof(b.ret)) +
+		int(unsafe.Sizeof(b.id)) +
+		int(unsafe.Sizeof(b.cpu)) +
+		int(unsafe.Sizeof(b.line)) +
+		int(len(b.file))
+}
+
+func (b *BPFError) Decode(rawBuffer []byte) error {
+	if len(rawBuffer) < b.Size() {
+		return fmt.Errorf("can't decode error raw data - buffer of %d should be of %d bytes", len(rawBuffer), b.Size())
+	}
+
+	b.count = binary.LittleEndian.Uint32(rawBuffer[0:4])
+	b.padding = binary.LittleEndian.Uint32(rawBuffer[4:8])
+
+	// bpf_error
+	b.ret = int64(binary.LittleEndian.Uint64(rawBuffer[8:16]))
+	b.id = BPFErrorType(binary.LittleEndian.Uint32(rawBuffer[16:20]))
+	b.cpu = binary.LittleEndian.Uint32(rawBuffer[20:24])
+	b.line = binary.LittleEndian.Uint32(rawBuffer[24:28])
+	copy(b.file[:], rawBuffer[28:104])
+	// bpf_error
+
+	return nil
+}
+
+func (t *Tracee) processBPFErrors() {
+	t.stats.BPFErrorsCount = make(counter.CountersMap)
+
+	for {
+		select {
+		case rawData := <-t.bpfErrorsChannel:
+			if len(rawData) == 0 {
+				continue
+			}
+
+			bpfErr := BPFError{}
+			if err := bpfErr.Decode(rawData); err != nil {
+				t.handleError(fmt.Errorf("consumeBPFErrorsChannel: decode - %v", err))
+				continue
+			}
+			// This logger timestamp in no way reflects the ebpf error original time
+			// since bpf errors channel is populated with aggregated logs
+			logger.Error(bpfErr.Error(),
+				"id", bpfErr.ID(),
+				"type", bpfErr.Type().String(),
+				"ret", bpfErr.Return(),
+				"cpu", bpfErr.CPU(),
+				"file", bpfErr.FileAsString(),
+				"line", bpfErr.Line(),
+				"count", bpfErr.Count(),
+			)
+			// Create a minor key to keep track of the bpf error origin,
+			// as the BPFError type has count which is volatile
+			orig := getBPFErrorOrigin(bpfErr)
+			if cnt, found := t.stats.BPFErrorsCount[orig]; !found {
+				t.stats.BPFErrorsCount[orig] = counter.NewCounter(uint64(bpfErr.Count()))
+			} else {
+				cnt.Set(uint64(bpfErr.Count()))
+			}
```

Comment:
@yanivagman I returned to my previous understanding. The perfbuf is aggregated with origin (which already contains cpu) and its count. To correctly set the count, we must track the origin. If we do it percpu only we'll override count when errors arrive from different origins (ret/id/cpu/line/file). So this CounterMap comes as a solution.

Reply:
After further analysis I decided to simplify this by [resetting](https://github.com/aquasecurity/tracee/pull/2352/files#diff-65cb2098ddddc611ef7711cd72ebd7fd5f6bb0e9bfddef3b5600305e8bdf3f80R1083) the count in the ebpf after submission. In the future if we need to keep percpu counts in user space, we can revisit this for improvement.